athena-elasticsearch/src/main/java/com/amazonaws/athena/connectors/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -189,7 +191,16 @@ public ListTablesResponse doListTables(BlockAllocator allocator, ListTablesReque
         List<TableName> tableNames = Stream.concat(indicesStream, getDataStreamNames(client))
                 .map(tableName -> new TableName(request.getSchemaName(), tableName))
                 .collect(Collectors.toList());
-        return new ListTablesResponse(request.getCatalogName(), tableNames, null);
+
+        int startToken = request.getNextToken() == null ? 0 : Integer.parseInt(request.getNextToken());
+        String nextToken = request.getPageSize() != UNLIMITED_PAGE_SIZE_VALUE ? Integer.toString(startToken + request.getPageSize()) : null;
+
+        if (request.getPageSize() != UNLIMITED_PAGE_SIZE_VALUE) {
+            tableNames = tableNames.stream().skip(startToken).limit(request.getPageSize()).collect(Collectors.toList());
+            logger.info("{} tables returned from start token, {}, to end token, {}", tableNames.size(), startToken, nextToken);
+        }
+
+        return new ListTablesResponse(request.getCatalogName(), tableNames, nextToken);
```

Comment:
You can delay the materialization of the stream since you don't actually need it as a list yet.

Meaning lines 191 to 203 can look like this instead:

```
        List<TableName> tableNamesStream = Stream.concat(indicesStream, getDataStreamNames(client))
                .map(tableName -> new TableName(request.getSchemaName(), tableName));

        int startToken = request.getNextToken() == null ? 0 : Integer.parseInt(request.getNextToken());
        String nextToken = request.getPageSize() != UNLIMITED_PAGE_SIZE_VALUE ? Integer.toString(startToken + request.getPageSize()) : null;

        if (request.getPageSize() != UNLIMITED_PAGE_SIZE_VALUE) {
            tableNamesStream = tableNamesStream.skip(startToken).limit(request.getPageSize());
        }

        List<TableName> tableNames = tableNamesStream.collect(Collectors.toList());
        logger.info("{} tables returned from start token, {}, to end token, {}", tableNames.size(), startToken, nextToken);

        return new ListTablesResponse(request.getCatalogName(), tableNames, nextToken);
```

Additionally, you will want to check if `indicesStream` and `getDataStreamNames()` are actually ordered streams.